source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -79,34 +79,50 @@ void DnsCacheImpl::startCacheLoad(const std::string& host, uint16_t default_port
     return;
   }
 
-  // TODO(mattklein123): Figure out if we want to support addresses of the form <IP>:<port>. This
-  //                     seems unlikely to be useful in TLS scenarios, but it is technically
-  //                     supported. We might want to block this form for now.
-  const auto colon_pos = host.find(':');
+  // First try to see if there is a port included. This also checks to see that there is not a ']'
+  // as the last character which is indicative of an IPv6 address without a port. This is a best
+  // effort attempt.
+  const auto colon_pos = host.rfind(':');
   absl::string_view host_to_resolve = host;
-  if (colon_pos != absl::string_view::npos) {
+  if (colon_pos != absl::string_view::npos && host_to_resolve.back() != ']') {
     const absl::string_view string_view_host = host;
     host_to_resolve = string_view_host.substr(0, colon_pos);
     const auto port_str = string_view_host.substr(colon_pos + 1);
     uint64_t port64;
     if (port_str.empty() || !absl::SimpleAtoi(port_str, &port64) || port64 > 65535) {
       // Just attempt to resolve whatever we were given. This will very likely fail.
-      // TODO(mattklein123): Should we actually fail here or do something different?
       host_to_resolve = host;
     } else {
       default_port = port64;
     }
   }
 
+  // Now see if this is an IP address. We need to know this because some things (such as setting
+  // SNI) are special cased if this is an IP address. Either way, we still go through the normal
+  // resolver flow. We could short-circuit the DNS resolver in this case, but the extra code to do
+  // so is not worth it since the DNS resolver should handle it for us.
+  bool is_ip_address = false;
+  try {
+    absl::string_view potential_ip_address = host_to_resolve;
+    if (potential_ip_address.front() == '[' && potential_ip_address.back() == ']') {
```

Comment:
Do you think it's worth tweaking parseInternetAddress to handle this?
I don't know if it's the only case where we'll have the [] without a port.

Reply:
I could go either way on this. AFAICT, `[]` is only required when embedded in a host/authority header per the spec, but happy to do whatever the group thinks is best. @lizan @PiotrSikora?

Reply:
[RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2) requires `[]` for URI too, and actually `parseInternetAddressAndPort` already does sth similar to here. So maybe refactor with that too? It can be a TODO.

Reply:
OK, I will add a relevant TODO here.

Reply:
I just looked at this and I realized that I still need the trimmed address for sending into the DNS resolver, since I don't believe that code handles the brackets, so I would rather leave this as is since I would still need to do it in both places. I can still add a TODO to try to figure out something better. Is that OK?

Reply:
Sounds reasonable to me.